### ISSUE TITLE
Blocking eglot-imenu

### ISSUE THREAD

**Issue by phst** [joaotavora/eglot#212]
Currently `eglot-imenu` is blocking and waits using the default timeout. However, it's called in places where blocking calls are extremely undesirable, e.g. in the idle timer used by `which-function-mode`. This makes buffers unusable if `which-function-mode` is enabled and the server doesn't react immediately. I'd suggest to make `eglot-imenu` asynchronous, i.e. it should return a cached result and fire off an asynchronous update in the background. That means the results might be a bit stale, but that's still better than blocking input.

**Comment by phst** [comment]
OK, I'll try to come up with something basic.

**Comment by ambihelical** [comment]
Oh it seems this bug is why my editting keystrokes have been getting eaten randomly (mainly with rust-mode for some reason), this was driving me up the wall.  Turning off which-key when I'm editing seems to be a short-term workaround, at least it reduces the severity of the problem.

**Comment by alexanderbarbosa** [comment]
@ambihelical what is the relation of which-function and which-key, I dont get it...

**Comment by ambihelical** [comment]
@alexanderbarbosa My mistake. I'm probably trolling myself thinking turning off which-key helped, although my keystrokes do seem to be eaten less to some degree.

**Comment by joaotavora** [comment]
@phst I had a nother look at this PR, and it sounds like something that `which-function-mode` should take care of.  I.e. it should not be the particular implementation of `imenu`'s interfaces that should be worried about this.  Also, I tried `which-func-mode` and either it is _not_ using `imenu` at all, or it is already using that caching behaviour.

So I am closing this, but will reopen if you insist.

**Comment by xwl** [comment]
I'm having the same issue as the @phst reported. It is barely usable for me when which-function-mode is enabled. By checking the backtrace with C-g, it is indeed blocked by the call eglot-imenu.

What would be proper solution here? It sounds reasonable to me, that we make elgot-imenu return a cached value right away, then updating the result in the background.

**Comment by joaotavora** [comment]
Pinging @monnier since he's probably interested in this, too.

> What would be proper solution here? It sounds reasonable to me, that we make elgot-imenu return a cached value right away, then updating the result in the background.

This problem is complicated because `imenu` is meant to be used synchronously and, implicitly through `which-func-mode`, asynchronously.  But to bother the backends such as `eglot-imenu` about that is a bad idea.  To have them work around it in caches and their ever-complicated cache-invalidation semantics is the the kind of fix-it-for-me-now misdesign that leads to much spaghetti, more than often found in ElDoc and Company backends, for example.

Designing it well means for this things to develop a good async API, like ElDoc recently has.  The problem in `which-func-mode` is slightly complicated by the fact that it, by default, wants to use `imenu`, which is synchronous by default.

So the next best thing is to _do_ bother the backend, but just a little bit. These frontends, like `which-func-mode` should bind the variable `non-essential` to `t` around their non-essential but nice-to-have calls to backends. For example `company-capf` binds it and thus can work with a `eglot-completion-at-point` which is synchronous and blocking by default (because it _needs_ to be given current APIs), but knows how to give up if the user is typing stuff that invalidates the previous request.

To summarize, if `which-func-mode` starts binding `non-essential` the problem will likely be solved easily in Eglot by this one-liner:

```diff
diff --git a/eglot.el b/eglot.el
index 35c959b..4ad8c8f 100644
--- a/eglot.el
+++ b/eglot.el
@@ -2414,7 +2414,8 @@ is not active."
               (jsonrpc-request (eglot--current-server-or-lose)
                                :textDocument/documentSymbol
                                `(:textDocument
-                                 ,(eglot--TextDocumentIdentifier))))))))
+                                 ,(eglot--TextDocumentIdentifier))
+                               :cancel-on-input non-essential))))))
 
 (defun eglot--apply-text-edits (edits &optional version)
   "Apply EDITS for current buffer if at VERSION, or if it's nil."
```

If `which-func.el` is _not_ binding that, then I think it should start (dunno, maybe _every_ idle timer should bind it by default).   In that case what you @xwl and @phst can do is investigate in Emacs devel if it is possible to update `which-func.el` with this change, and furthermore, if one can make `which-func.el` a "core ELPA package", so that the results become useful for people working with older versions of Emacs.

**Comment by xwl** [comment]
Thanks for the detailed explanation. You may be very right, that which-func should not use imenu in the first place. I have just noticed that even inside the which-func.el, it has a TODO for removing dependency of imenu. If that TODO is finished, then no changes is needed at all here in eglot.

    ;; TODO LIST
    ;; ---------
    ;;     1. Dependence on imenu package should be removed.  Separate
    ;; function determination mechanism should be used to determine the end
    ;; of a function as well as the beginning of a function.
    ;;     2. This package should be realized with the help of overlay
    ;; properties instead of imenu--index-alist variable.

**Comment by joaotavora** [comment]
> You may be very right, that which-func should not use imenu in the first place

Sure, the problem is that's a tall order. Because you want current code to work and this approach is the status quo and works most of the time.  Maybe `which-func` could use just `beginning-of-defun`  and `end-of-defun`.  But not all major modes implement that. You'd need a backward compatible solution and maybe fall back on imenu.

Anyway, I really encourage to work on it.  To solve _this_ immediate problem, I'd go for the `:cancel-on-input` trick, tho.

**Comment by monnier** [comment]
> This problem is complicated because `imenu` is meant to be used
> synchronously and, implicitly through `which-func-mode`,
> asynchronously.

Indeed, `imenu` and `which-func-mode` both use quite naive APIs designed
decades ago and which are in need of some love (and considering the
`C-x 4 a` case along with it).

> To summarize, if `which-func-mode` starts binding `non-essential` the
> problem will likely be solved easily in Eglot by this one-liner:

I just pushed a patch to Emacs's `master` that does that.

Note that not all uses of which-func are asynchronous (which-func can
also be used by `C-x 4 a`).


        Stefan

**Comment by joaotavora** [comment]
> I just pushed a patch to Emacs's `master` that does that.

And do you think you could release it as a `:core` ELPA package, maybe? Or is it too tangled?

**Comment by monnier** [comment]
>> I just pushed a patch to Emacs's `master` that does that.
> And do you think you could release it as a `:core` ELPA package, maybe? Or is it too tangled?

No idea, feel free to look into that.

**Comment by joaotavora** [comment]
Nice dodge ahaha

Thanks anyway, Stefan
